File: cpp/q004.cpp
#include <iostream>
#include <string>
#include <cmath>
using namespace std;
int palindrome(int number){
    int n = number;
    int reversed = 0;
    int digit;
    while (n > 0) {
        digit = n % 10;
        reversed = reversed * 10 + digit;
        n /= 10;
    }
    return reversed;
}

int main(){
    int ans=0;
    for (int i=100; i<1000; i++){
        for (int j = 100; j<1000;j++){
            if (i*j == palindrome(i*j)){
                ans = max(ans, i*j);
            }
        }
    }
    cout << ans << endl;
    return 0;
}

File: rust/q004.rs

fn palindrome(number: u64) -> u64 {
    let mut n = number;
    let mut reversed = 0;
    while n > 0 {
        let digit = n % 10;
        reversed = reversed * 10 + digit;
        n /= 10;
    }
    reversed
}

fn main(){
    let mut ans = 0;
    for i in 100..1000{
        for j in 100..1000{
            let product = i*j;
            let pal = palindrome(product);
            if product == pal && pal > ans{
                ans = pal;
            }
        }
    }
    println!("{}", ans)
}